orbit/pkg/installer/installer.go

```diff
@@ -0,0 +1,251 @@
+	"github.com/fleetdm/fleet/v4/pkg/file"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+	"github.com/osquery/osquery-go"
+	osquery_gen "github.com/osquery/osquery-go/gen/osquery"
+	"github.com/rs/zerolog/log"
+)
+
+type QueryResponse = osquery_gen.ExtensionResponse
+type QueryResponseStatus = osquery_gen.ExtensionStatus
+
+// Client defines the methods required for the API requests to the server. The
+// fleet.OrbitClient type satisfies this interface.
+type Client interface {
+	GetInstallerDetails(installId string) (*fleet.SoftwareInstallDetails, error)
+	DownloadSoftwareInstaller(installerID uint, downloadDir string) (string, error)
+	SaveInstallerResult(payload *fleet.HostSoftwareInstallResultPayload) error
+}
+
+type QueryClient interface {
+	Query(context.Context, string) (*QueryResponse, error)
+}
+
+type Runner struct {
+	OsqueryClient QueryClient
+	OrbitClient   Client
+
+	// osquerySocketPaht is used to establish the osquery connection
+	// if it's ever lost or disconnected
+	osquerySocketPaht string
+
+	// tempDirFn is the function to call to get the temporary directory to use,
+	// inside of which the script-specific subdirectories will be created. If nil,
+	// the user's temp dir will be used (can be set to t.TempDir in tests).
+	tempDirFn func(dir, pattern string) (string, error)
+
+	// execCmdFn can be set for tests to mock actual execution of the script. If
+	// nil, execCmd will be used, which has a different implementation on Windows
+	// and non-Windows platforms.
+	execCmdFn func(ctx context.Context, scriptPath string, env []string) ([]byte, int, error)
+
+	// can be set for tests to replace os.RemoveAll, which is called to remove
+	// the script's temporary directory after execution.
+	removeAllFn func(string) error
+
+	connectOsquery func(*Runner) error
+
+	scriptsEnabled func() bool
+}
+
+func NewRunner(client Client, socketPath string, scriptsEnabled func() bool) *Runner {
+	r := &Runner{
+		OrbitClient:       client,
+		osquerySocketPaht: socketPath,
+		scriptsEnabled:    scriptsEnabled,
+	}
+
+	return r
+}
+
+func (r *Runner) Run(config *fleet.OrbitConfig) error {
+	if r.connectOsquery == nil {
+		r.connectOsquery = connectOsquery
+	}
+
+	if err := r.connectOsquery(r); err != nil {
+		return fmt.Errorf("software installer runner connecting to osquery: %w", err)
+	}
+	return r.run(context.Background(), config)
+}
+
+func connectOsquery(r *Runner) error {
+	if r.OsqueryClient == nil {
+		osqueryClient, err := osquery.NewClient(r.osquerySocketPaht, 10*time.Second)
+		if err != nil {
+			log.Err(err).Msg("establishing osquery connection for software install runner")
+			return err
+		}
+
+		r.OsqueryClient = osqueryClient.Client
+	}
+
+	return nil
+}
+
+func (r *Runner) run(ctx context.Context, config *fleet.OrbitConfig) error {
+	var errs []error
+	for _, installerID := range config.Notifications.PendingSoftwareInstallerIDs {
+		if ctx.Err() == nil {
```

Comment:
Seems like we otherwise ignore the context error, I think we should so something like this?
```suggestion
		if err := ctx.Err(); err != nil {
		    return err
		}
```
(or append to `errs` and break out of the loop)

Reply:
True! I will append it to the errors and return instead

Reply:
Done!